### PR TITLE
Feature/wd 2077/bankid

### PR DESCRIPTION
**Pull request by boscar**
Implement bankid support.

### CODE REVIEW

File: src/bankid/message-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+const userSign = () =>
+    "Skriv in din säkerhetskod i BankID-appen och välj Legitimera eller Skriv under.";
+
+const expiredTransaction = () =>
+    "BankID-appen svarar inte. Kontrollera att den är startad och att du har internetanslutning. Om du inte har något giltigt BankID kan du hämta ett hos din Bank. Försök sedan igen.";
+
+const certificateErr = () =>
+    "Det BankID du försöker använda är för gammalt eller spärrat. Använd ett annat BankID eller hämta ett nytt hos din internetbank.";
+
+const userCancel = () => "Åtgärden avbruten.";
```
Comment (ourbjorn) [inline src/bankid/message-resolver.ts:24]: The diacritic above the uppercase A seems to be in some odd encoding?

File: src/bankid/index.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    return api
+        .collect(request)
+        .then(response => new BankIdCollectResponse(response, request.method));
+};
+
+export const cancel = (request: IBankIdCancelRequest) => {
+    BankIdCancelRequestBuilder.validate(request);
+
+    return api.cancel(request).then(response => response);
```
Comment (ourbjorn) [inline src/bankid/index.ts:39]: Isn't `.then(response => response)` excessive here?

File: src/http/apis/bankid/request-builder.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { IBankIdAuthRequest } from "../../../bankid/types";
```
Comment (ourbjorn) [inline src/http/apis/bankid/request-builder.ts:1]: Good move, moving these types of functions into their own context/file!

File: src/http/apis/bankid/utils.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import Configuration from "../../../config/index";
```
Comment (ourbjorn) [inline src/http/apis/bankid/utils.ts:1]: I don't think these utility functions create enough abstractions for what they do. Having the logic closer to the actual API-functionality renders, IMO, more readable code and code that's easier to maintain.

File: src/index.ts [hunk 5]
```diff
@@ -63,3 +65,9 @@ export {
     IVehicleLookupResponse,
     VehicleCondition,
 } from "./vehicles/types";
+export {
```
Comment (ourbjorn) [inline src/index.ts:68]: It feels like we're not exporting enough types here - shouldn't we export the `*Request`-types here at least?

File: src/bankid/message-resolver.ts (hunk 1 above)
Comment (boscar) [reply src/bankid/message-resolver.ts:24]: Nicely spotted.

File: src/bankid/index.ts (hunk 2 above)
Comment (boscar) [reply src/bankid/index.ts:39]: sure is.

File: src/http/apis/bankid/request-builder.ts (hunk 3 above)
Comment (boscar) [reply src/http/apis/bankid/request-builder.ts:1]: thanks.

File: src/index.ts (hunk 5 above)
Comment (boscar) [reply src/index.ts:68]: We could, but it is not strictly required when we export the builder. But we probably should, I will change that.

File: src/http/apis/bankid/utils.ts (hunk 4 above)
Comment (boscar) [reply src/http/apis/bankid/utils.ts:1]: I see what you mean, although I don't want to keep all api logic for bank id in a single file.

After thinking about it I think that the best way would be to separate all api logic for bank id into three files: auth, collect and cancel. Each of these files include their own separate _util functions_.

**Pull request closed without merging**